multipart-post.lua

```diff
@@ -36,44 +40,111 @@ local gen_boundary = function()
+    if _t == "string" then
+        section_header(r, k, {})
+    elseif _t == "table" then
+        assert(v.data, "invalid input")
+        local extra = {
+            filename = v.filename or v.name,
+            content_type = v.content_type or v.mimetype
+                or "application/octet-stream",
+            content_transfer_encoding = v.content_transfer_encoding or "binary",
+        }
+        section_header(r, k, extra)
+    else
+        error(string.format("unexpected type %s", _t))
+    end
+end
+
+local encode_source = function(k, v, boundary)
+    r = {}
+    encode(r, k, v, boundary)
+    return ltn12.source.string(table.concat(r))
+end
+
+local data_len = function(d)
+    local _t = type(d)
+
+    if _t == "string" then
+        return string.len(d)
+    elseif _t == "table" then
+        if type(d.data) == "string" then
+            return string.len(d.data)
+        end
+        if d.len then return d.len end
+        error("must provide data length for non-string datatypes")
+    end
+end
+
+local content_length = function(t, boundary)
     local r = {}
-    local _t
-    for k,v in pairs(t) do
-        tprintf(r, "--%s\r\n", boundary)
-        _t = type(v)
+    local data_length = 0
+    for k, v in pairs(t) do
+        encode(r, k, v, boundary)
+        data_length = data_length + data_len(v)
+        tprintf(r, "\r\n")
+    end
+    tprintf(r, "--%s--\r\n", boundary)
+
+    return string.len(table.concat(r)) + data_length
+end
+
+local get_data_src = function(v)
+    local _t = type(v)
+    if v.source then
+        return v.source
+    elseif _t == "string" then
+        return ltn12.source.string(v)
+    elseif _t == "table" then
+        _t = type(v.data)
         if _t == "string" then
-            append_data(r, k, v, {})
+            return ltn12.source.string(v.data)
         elseif _t == "table" then
-            assert(v.data, "invalid input")
-            local extra = {
-                filename = v.filename or v.name,
-                content_type = v.content_type or v.mimetype
-                    or "application/octet-stream",
-                content_transfer_encoding = v.content_transfer_encoding or "binary",
-            }
-            append_data(r, k, v.data, extra)
-        else error(string.format("unexpected type %s", _t)) end
+            return ltn12.source.table(v.data)
+        elseif _t == "userdata" then
+            return ltn12.source.file(v.data)
+        elseif _t == "function" then
+            return v.data
+        end
     end
-    tprintf(r, "--%s--\r\n", boundary)
-    return table.concat(r), boundary
+    error("invalid input")
+end
+
+local set_ltn12_blksz = function(sz)
+    assert(type(sz) == "number", "set_ltn12_blksz expects a number")
+    ltn12.BLOCKSIZE = sz
+end
+_M.set_ltn12_blksz = set_ltn12_blksz
+
+local source = function(t, boundary)
+    local n = 1
+    local sources = {}
+    for k, v in pairs(t) do
+        sources[n] = encode_source(k, t[k], boundary)
```

Comment:
detail: `t[k]` is `v`

Also the naming of `encode` and `encode_source` is not precise enough, it should be something like `encode_header_as_string` and `encode_header_as_source` or something like that.